protocol/0047-DSRF-data_source_filter.md

```diff
@@ -60,6 +60,8 @@ To specify a filtered data source the following parameters can be specified:
 
 - `data`: (required) another data source definition defining the input data
 - `filters`: (required) a list of _at least one_ filter to apply to the data
+- `data_if_true`: (optional) any data source, if specified the latest value received from that downstream data source gets returned from the current data source whenever the result of applying `filters` to `data` is not null, if downstream data source has not been specified then the result of applying `filters` to `data` gets returned directly.
+- `data_if_false`: (optional) any data source, if specified the latest value received from that downstream data source gets returned from the current data source whenever the result of applying `filters` to `data` is null, if downstream data source has not been specified then nothing gets returned.
```

Comment:
@vegaprotocol/core please note that this is what seemed to be the easiest change from spec point of view to achieve the following functionality:
* the marke settles at `max price` if `settlement price` >= `max price`,
* the market settles at `0` if `settlement price < max price`.

But I don't think this spec accurately represents the implementation so definitely open to suggestions on how to achieve the functionality specified above with least amount of core work while maintaining a sensible UX.

Reply:
I don't think doing it this would would be much core work, maybe how we represent that this happened in the data-node would probably be the more fiddly bit. I might suggest changing the names to `override_if_XXX` instead.

I also can't help feel that this logic should be in the product code though? So the product gets the data-source as usual and then the code in the product just compares the settlement price in the data-source to `max price` and makes the decision itself. Because doing that would just be a couple of lines of Go and avoids adding complexity to the data-source definitions.

Reply:
It's more that we wanted to both allow experimentaion (have a capped future that behaves exactly as a regular fixed-expiry future, but all prices are capped) as well as be able to reproduce binary options markets: the market settles either at `0` or `max_price` and nothing in between.

Reply:
Though I guess something like `allow_intermediate_settlement_price` flag in the product itself would do it.

Reply:
I am not super comfortable having this outside data sourcing because all views and APIs will report the data source value as the settlement price. Any tooling, fornt ends, etc. would have to special-case this. The settlement data source should match the price used to settle a future or perp IMO.

Reply:
Any external oracle. I think this needs to be covered within the system so parties don't have to take on a risk of a payout they aren't expecting (so they don't need to trust that the oracle emits either `0` or `max-price`, but it's guaranteed by the product).

Reply:
> What are the expected data-sources for this binary markets, 

Any existing DEX on Ethereum/Gnosis/Arbitrum, price oracles like Pyth, and truth oracles like UMA. We would expect to be able to settle directly off the data and also off the answers to questions about the current and historical data.  

> and is there a reason can't just emit the data in a form that it is expected to be in i.e 0 or max-price?

I think this is actually the right question. We have the Eth oracle framework so data sourcing logic that ***should*** be in Vega can be built into wrapper contracts until we fix data sourcing properly.

So I guess the `allow_intermediate_settlement_price` flag will do fine. 

I am not so sure about making any intermediate number 0 otherwise, not about capping the settlement price to MAX. I am trying to to come up with a reason not to rejected invalid settlement data that's bigger than MAX, or between 0 and MAX if the intermediate price flag is false rather than changing the value?

Reply:
My motivation was so that it's easy to setup markets "like will BTC a week from now be >70k" without any extra setup using fairly simple oracles that emit a price (something that we'd use for settling a vanilla fixed-expiry futures market).